peer/peerdiscovery.go

```diff
@@ -0,0 +1,134 @@
+package peer
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"github.com/renproject/surge"
+	"time"
+
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	"go.uber.org/zap"
+)
+
+type DiscoveryClient struct {
+	opts DiscoveryOptions
+
+	transport *transport.Transport
+}
+
+func NewDiscoveryClient(opts DiscoveryOptions, transport *transport.Transport) *DiscoveryClient {
+	return &DiscoveryClient{
+		opts: opts,
+
+		transport: transport,
+	}
+}
+
+func (dc *DiscoveryClient) DiscoverPeers(ctx context.Context) {
+	var pingData [2]byte
+	binary.LittleEndian.PutUint16(pingData[:], dc.transport.Port())
+
+	msg := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
+	defer ticker.Stop()
+	for {
+		for _, sig := range dc.transport.Table().Peers(dc.opts.Alpha) {
+			msg.To = id.Hash(sig)
+			err := dc.transport.Send(ctx, sig, msg)
```

Comment:
This `ctx` should be scoped. That is, we should create a wrapping context that has a smaller timeout, because `ctx` as defined at this point is the _entire lifetime_ of the peer discovery loop. This is typically going to be the entire lifetime of the application. Obviously way too high for a single request.